pkg/services/authn/clients/ext_jwt.go

```diff
@@ -46,68 +55,97 @@ type ExtendedJWT struct {
 	signingKeys signingkeys.Service
+	verifier    authlib.Verifier[ExtendedJWTClaims]
 }
 
 type ExtendedJWTClaims struct {
 	jwt.Claims
-	ClientID     string              `json:"client_id"`
-	Groups       []string            `json:"groups"`
-	Email        string              `json:"email"`
-	Name         string              `json:"name"`
-	Login        string              `json:"login"`
-	Scopes       []string            `json:"scope"`
-	Entitlements map[string][]string `json:"entitlements"`
+	// Access policy scopes
+	Scopes []string `json:"scopes"`
+	// Grafana roles
+	Permissions []string `json:"permissions"`
+	// On-behalf-of user
+	DelegatedPermissions []string `json:"delegatedPermissions"`
 }
 
 func (s *ExtendedJWT) Authenticate(ctx context.Context, r *authn.Request) (*authn.Identity, error) {
-	jwtToken := s.retrieveToken(r.HTTPRequest)
+	jwtToken := s.retrieveAuthenticationToken(r.HTTPRequest)
 
-	claims, err := s.verifyRFC9068Token(ctx, jwtToken)
+	claims, err := s.verifyRFC9068Token(ctx, jwtToken, rfc9068ShortMediaType)
 	if err != nil {
 		s.log.Error("Failed to verify JWT", "error", err)
 		return nil, errJWTInvalid.Errorf("Failed to verify JWT: %w", err)
 	}
 
-	// user:id:18
-	userID, err := strconv.ParseInt(strings.TrimPrefix(claims.Subject, fmt.Sprintf("%s:id:", authn.NamespaceUser)), 10, 64)
-	if err != nil {
-		s.log.Error("Failed to parse sub", "error", err)
-		return nil, errJWTInvalid.Errorf("Failed to parse sub: %w", err)
-	}
+	idToken := s.retrieveAuthorizationToken(r.HTTPRequest)
+	if idToken != "" {
+		idTokenClaims, err := s.verifyRFC9068Token(ctx, idToken, "jwt")
+		if err != nil {
+			s.log.Error("Failed to verify id token", "error", err)
+			return nil, errJWTInvalid.Errorf("Failed to verify id token: %w", err)
+		}
 
-	// FIXME: support multiple organizations
-	defaultOrgID := s.getDefaultOrgID()
-	if r.OrgID != defaultOrgID {
-		s.log.Error("Failed to verify the Organization: OrgID is not the default")
-		return nil, errJWTInvalid.Errorf("Failed to verify the Organization. Only the default org is supported")
+		return s.authenticateAsUser(idTokenClaims, claims)
 	}
 
-	signedInUser, err := s.userService.GetSignedInUserWithCacheCtx(ctx, &user.GetSignedInUserQuery{OrgID: defaultOrgID, UserID: userID})
-	if err != nil {
-		s.log.Error("Failed to get user", "error", err)
-		return nil, errJWTInvalid.Errorf("Failed to get user: %w", err)
-	}
+	return s.authenticateAsService(claims)
+}
 
-	if signedInUser.Permissions == nil {
-		signedInUser.Permissions = make(map[int64]map[string][]string)
+func (s *ExtendedJWT) authenticateAsUser(idTokenClaims,
+	accessTokenClaims *ExtendedJWTClaims) (*authn.Identity, error) {
+	// Only allow access policies to impersonate
+	if !strings.HasPrefix(accessTokenClaims.Subject, fmt.Sprintf("%s:", authn.NamespaceAccessPolicy)) {
+		s.log.Error("Invalid subject", "subject", accessTokenClaims.Subject)
+		return nil, errJWTInvalid.Errorf("Failed to parse sub: %s", "invalid subject format")
```

Comment:
we should move this check to Authenticate and remove from `authenticateAsUser` and `authenticateAsService`